Client: move `request_queue` outside `vsr.Client`

kprotty opened the pull request:
First step to making tb_client packets intrusive is having each packet map 1:1 to a `vsr.Client.Request`. This was awkward to do as `vsr.Client` is specific to `StateMachine` and holds a callback but `Packet` is type-erased with `tb_client_t` holding the callback..

An observation was made that most uses of `vsr.Client` are sequential and don't actually need a queue of concurrent requests (aof, benchmark, demo, repl). The only uses which queue are `tb_client` requiring batching/demuxing and `Cluster/Simulator` requiring specific queue semantics. So lets make `vsr.Client` only handle one message at a time, delegating queueing to its caller. Apart from simplifying internals, this means Packet (which `tb_client` already queues up outside VSR) now implicitly maps to a VSR request to be intrusive later.

Batching/demuxing was moved directly into `tb_client`, necessitating `Packet` to be a bit thicker. I bumped its size from 32b to 64b with an extra `reserved: [2]usize` to keep the cache-line idea from TIGER_STYLE. The thread-safe submission stack was also moved into `Context` so that requeuing could push to an unsynchronized FIFO instead and not call `signal.notify()` spuriously. This improved batched request throughput in language client tests.

sentientwaffle commented on src/testing/cluster.zig:
```diff
@@ -61,6 +62,10 @@ pub fn ClusterType(comptime StateMachineType: anytype) type {
 
+        const ClientQueue = RingBuffer(*Message.Request, .{
+            .array = constants.client_request_queue_max,
+        });
```
Is this just so that `Cluster.request()`'s API stays the same? It might make more sense to modify the state machine workload so that at most one request per client is ever in flight.

kprotty committed:
client: serial vsr.Client and batching in tb_client

kprotty committed:
client: rebuild C bindings

kprotty committed:
simulator: mismatching Message type typo

kprotty committed:
client: remove request check & fix packet usage

raw_request() is the one to set Header.request so it shouldnt
check it before-hand as it's likely undefined.

kprotty committed:
client: vsr.Client based DemuxerType

The EchoClient used the default StateMachine.DemuxerType which
expects the bytes to be Results instead of the echoed back Events.

This fix allows the client to override the DemuxerType functionality
so that vsr.Client wraps the StateMachine's and EchoClient returns
back unmodified Event slices.

kprotty committed:
client: unit tests for EchoClient.DemuxerType

kprotty committed:
echo_client: tidy + fmt

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -170,12 +101,11 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
-        /// A client is allowed at most one inflight request at a time at the protocol layer.
-        /// We therefore queue any further concurrent requests made by the application layer.
-        request_queue: RequestQueue = RequestQueue.init(),
+        /// Tracks a currently processing request message submitted by `raw_request`.
```
```suggestion
        /// Tracks a currently processing (non-register) request message submitted by `raw_request`.
```

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -30,98 +29,30 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
-                assert(message.header.command == .request);
+        /// Custom Demuxer which passes through to StateMachine.Demuxer.
+        pub fn DemuxerType(comptime operation: StateMachine.Operation) type {
+            return struct {
```
Should this definition move nearer to where it is used?

kprotty replied:
It's used outside of `vsr.Client` (in the unit test below and tb_client Context).

sentientwaffle replied:
So why not move this (and the corresponding unit test) move to tb_client Context's file? It seems weird to have it here since `vsr.Client` is now unaware of batching.

kprotty replied:
It's due to `EchoClient`: its Demuxer must return Events back but StateMachine.Demuxer adjusts & returns Results. Creating a stub StateMachine just for EchoClient, with most of the public decls from the original but a custom Demuxer, seemed more weird. So instead, `vsr.Client` and `EchoClient` each have their own Demuxer wrapper and corresponding unit test.

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -296,208 +221,81 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
-                .event_count = @intCast(@divExact(message.body().len, event_size)),
-                .event_offset = 0,
-            };
+            assert(!message.header.operation.vsr_reserved());
```
Does this assertion need to move to `request()`? `raw_request` replays prepares during AOF recovery, which can include reserved operations. (I realize that this is not a new assertion).

kprotty replied:
`request()` already knows the operation is a StateMachine / not vsr reserved, so I think this assert can just be deleted for AOF. 

But in `Client.on_reply` it either expects a `.register` or a `StateMachine.Operation` to pass to the request callback. Would need to know how AOF intends to produce/handle this.

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -587,42 +385,48 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+                Request{ .message = message, .user_data = 0, .callback = undefined }
+            else if (self.request_inflight) |inflight|
+                inflight
+            else {
```
I think we can add
```suggestion
            else {
                assert(reply.header.request < self.request_number);
```

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -587,42 +385,48 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
                 return;
+            };
+
+            if (reply.header.request < inflight.message.header.request) {
```
```suggestion
            if (reply.header.request < inflight.message.header.request) {
                assert(self.register_inflight == null);
```

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -587,42 +385,48 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
 
-            var inflight = self.request_queue.pop().?;
-            const inflight_request = inflight.message.header.request;
+            assert(reply.header.request_checksum == inflight.message.header.checksum);
```
```suggestion
            assert(reply.header.request == inflight.message.header.request);
            assert(reply.header.request_checksum == inflight.message.header.checksum);
```

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -587,42 +385,48 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
-            // Even though we release our reference to the message, we might have another one
-            // retained by the send queue in case of timeout.
-            maybe(inflight.message.references > 0);
             inflight.message = undefined;
```
Could be be a little less eager? e.g. move this directly before the `if (inflight_vsr_operation == .register) {` branch that is responsible for invoking the callbacks. That minimize the duration that the message is released but not nulled.

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -783,14 +549,9 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
-            assert(self.request_queue.tail_ptr() == request);
-            request.demux_queue.push(demux);
+            assert(self.request_inflight == null);
+            assert(self.register_inflight == null);
```
Can these move up to the top of the function (after `if (self.request_number > 0) return;`) to serve as preconditions?

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -835,7 +596,13 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
-            assert(self.request_queue.head_ptr().?.message == message);
+            if (self.register_inflight) |inflight| {
+                maybe(self.request_inflight != null);
+                assert(inflight == message);
```
Swap the order of these asserts so that they mirror the `else` branch's asserts.

sentientwaffle commented on src/testing/cluster/state_checker.zig:
```diff
@@ -178,17 +178,17 @@ pub fn StateCheckerType(comptime Client: type, comptime Replica: type) type {
+                assert(message.header.request == header_b.?.request);
+                assert(message.header.command == .request);
+                assert(message.header.operation == header_b.?.operation);
+                assert(message.header.size == header_b.?.size);
```
I think `request` is a better name for this, to distinguish it from the prepare(s).

sentientwaffle commented on src/simulator.zig:
```diff
@@ -817,13 +818,13 @@ pub const Simulator = struct {
             // since they may start one at any time.
             reserved += @intFromBool(c.session == 0);
             // Count the number of requests queued.
```
```suggestion
            // Count the number of non-register requests queued.
```

sentientwaffle commented on src/simulator.zig:
```diff
@@ -817,13 +818,13 @@ pub const Simulator = struct {
 
-        // Make sure that there is capacity in the client's request queue.
-        if (client.messages_available == 0) return;
+        // Make sure that the client is free to send a new request
```
```suggestion
        // Make sure that the client is ready to send a new request
```
nit

sentientwaffle commented on src/fifo.zig:
```diff
@@ -14,7 +14,7 @@ pub fn FIFO(comptime T: type) type {
         count: u64 = 0,
         // This should only be null if you're sure we'll never want to monitor `count`.
-        name: ?[]const u8,
+        name: ?[*:0]const u8,
```
Is this change necessary for type reasons?

kprotty replied:
Tried to put a FIFO in Packet for batching but `[]const u8` wasn't compatible with `extern struct`. Removed the FIFO as not enough space so this is indeed a rogue change.

sentientwaffle commented on src/clients/c/tb_client/context.zig:
```diff
@@ -243,52 +267,152 @@ pub fn ContextType(
+
+                    // Check for pending packets of the same operation which can be batched.
+                    if (root.operation != packet.operation) continue;
+                    if (@intFromPtr(root.batch_data) + packet.data_size > constants.message_body_size_max) continue;
```
I'm confused... Why use `batch_data` (which is a pointer) to store the length?

kprotty replied:
It's mostly to save space in packet, but will add the extra field to make it simpler.

sentientwaffle commented on src/clients/c/tb_client/context.zig:
```diff
@@ -243,52 +267,152 @@ pub fn ContextType(
-            // Submit the message for processing:
-            self.client.batch_submit(
+            assert(offset == batch_size);
+            self.client.raw_request(
```
Hmm, so we are stuck using `raw_request` b/c we need to be the ones to assemble the batch. That is unfortunate (iirc `raw_request` was originally just for AOF recovery) but I don't see a nice way around it. I guess we could change `Client.request()`'s API to take a `message` too, but (unlike `raw_request`) it could ignore the header.

sentientwaffle commented on src/clients/c/tb_client/context.zig:
```diff
@@ -243,52 +267,152 @@ pub fn ContextType(
 
-            assert(packet.operation == @intFromEnum(op));
-            self.on_complete(packet, results);
+            // Submit the next pending packet after having VSR completed this one.
```
```suggestion
            // Submit the next pending packet now that VSR has completed this one.
```

sentientwaffle commented on src/clients/c/tb_client/context.zig:
```diff
@@ -243,52 +267,152 @@ pub fn ContextType(
+
+            switch (op) {
+                inline else => |operation| {
+                    // Demuxer expects []u8 but VSR callback provides []const u8.
+                    // The bytes are known to come from a Message body that will be soon discarded
+                    // therefor it's safe to @constCast and potentially modify the data in-place.
+                    var demuxer = Client.DemuxerType(operation).init(@constCast(reply));
```
Where do we modify the data? Or is it just the hypothetical -- i.e., that we could modify it if we wanted to.

kprotty replied:
It's modified in the main [StateMachine's Demuxer](<https://github.com/tigerbeetle/tigerbeetle/blob/main/src/state_machine.zig#L154-L159>) to update the `CreateTError.index`s to be packet-relative after having batched them. Should vsr.Client.Request.Callback just switch to `[]u8` instead?

sentientwaffle replied:
Aha, thanks! Yeah, if we are treating it as mutable then we probably should change the type. And ideally we would be able to assert `message.references == 1`, though I'm not positive if that would hold.

sentientwaffle commented on src/clients/node/src/node.zig:
```diff
@@ -234,14 +234,12 @@ fn request(
+    packet.data = packet_data.ptr;
+    packet.data_size = @intCast(packet_data.len);
+    packet.next = null;
+    packet.status = .ok;
```
I prefer assigning with `packet.* = ...` even if some fields need to be repeated, since it makes it obvious that all of the fields have a value assigned (or explicitly set to `undefined`, as in the case of `reserved`).

sentientwaffle commented on src/vsr/client.zig:
Since there is no more client request queue, we can reduce the number of messages that the client allocates. In `message_pool.zig`, replace:

```
    sum += constants.client_request_queue_max; // Client.request_queue
```

with:
```
    sum += 1; // Client.register_inflight
    sum += 1; // Client.request_inflight
```

Relatedly, are `Client.messages_available` and `constants.client_request_queue_max` still needed for anything, or can they be removed?

kprotty replied:
`messages_available` can be easily removed, but would need to double check on removing `client_request_queue_max` correctly. Will do so in its own PR.

sentientwaffle reviewed (commented):
Mostly just minor questions or suggestions... this is an excellent simplification!

kprotty committed:
client: fix uses of vsr.Release

kprotty committed:
client: cleanups and lower `messages_max_client`

kprotty committed:
client: adjust client_requests_queue_max after messages_max_client update

kprotty committed:
client: remove `messages_available`

Now that vsr.Client only sends one message at a time, this is no longer needed.
To see if a client can send a request, check `client.request_inflight == null` instead.

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -296,210 +210,79 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+            assert(message.header.cluster == self.cluster);
+            assert(message.header.command == .request);
+            assert(message.header.size >= @sizeOf(Header));
+            assert(message.header.size <= constants.message_size_max);
```
```suggestion
            assert(message.header.size <= constants.message_size_max);
            assert(message.header.operation.valid(StateMachine));
```

kprotty committed:
client: update packet fields to make batching simpler

`batch_data` was contextually repurposed as either the total batched
message body size or a pointer to the last batched packet. This was a
bit confusing to keep track of so they've been separated into their own
fields.

sentientwaffle commented on src/clients/c/tb_client/context.zig:
```diff
@@ -243,52 +267,144 @@ pub fn ContextType(
+                .cluster = self.client.cluster,
+                .command = .request,
+                .operation = vsr.Operation.from(StateMachine, operation),
+                .size = @intCast(@sizeOf(vsr.Header) + packet.batch_size),
```
With `batch_size` a `u32`, is this `@intCast` needed?

kprotty committed:
client: fix packet initialization + raw_request asserts

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -296,210 +210,80 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+            assert(message.header.command == .request);
+            assert(message.header.size >= @sizeOf(Header));
+            assert(message.header.size <= constants.message_size_max);
+            assert(message.header.operation.valid(StateMachine));
```
Can we assert that `message.header.timestamp == 0 or constants.aof_recovery);`?

kprotty replied:
This [was reached](<https://github.com/tigerbeetle/tigerbeetle/actions/runs/8364883560/job/22901324354?pr=1640>) under the existing `test_aof` step so I'll comment it out for now.

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -296,210 +210,80 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+            assert(message.header.command == .request);
+            assert(message.header.size >= @sizeOf(Header));
+            assert(message.header.size <= constants.message_size_max);
+            assert(message.header.operation.valid(StateMachine));
```
Also assert that `view`/`parent`/`session`/`request` are all zero.

kprotty committed:
client: fixup asserts

batiati commented on src/vsr/client.zig:
```diff
@@ -589,42 +372,46 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
-                }
-                assert(reply.header.request_checksum == inflight.message.header.checksum);
-            } else {
+            var inflight = if (self.register_inflight) |message|
+                Request{ .message = message, .user_data = 0, .callback = undefined }
+            else if (self.request_inflight) |inflight|
+                inflight
```
Is it worth documenting/asserting that `register_inflight`, when != null, is the equivalent of the old request_queue's head?

kprotty committed:
client: clarifying doc comments + disable aof assert

kprotty committed:
Merge branch 'main' into king/client-no-queue

sentientwaffle commented on src/clients/c/tb_client/echo_client.zig:
```diff
@@ -17,30 +18,43 @@ pub fn EchoClient(comptime StateMachine_: type, comptime MessageBus: type) type
+        pub const StateMachine = VSRClient.StateMachine;
+        pub const Request = VSRClient.Request;
+
+        /// Custom Demuxer which treats Event(operation)'s and results and echoes them back.
```
Does this mean to say s/"and results"/"as results"?

kprotty committed:
client: fix comment typo/todo
